gradle: add module

britter opened the pull request:
### Description

Introduces a new program called gradle for managing files stored in the
home directory by the [Gradle Build Tool](https://gradle.org). Gradle
uses the $HOME/.gradle folder for all it's configuration.

Features of the new program module are:

- Automatically setting programs.java.enable = true to make a Java
  installation available for running Gradle.
- Specifying an alternate Gradle home directory
- Setting of abitrary values for gradle.properties stored inside the
  Gradle home directory.
- Defining init scripts either by pointing to a directory that will be
  linked to init.d inside the Gradle home directory or by specifying
  a list of init script definitions.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.
- [x] Code formatted with `./format`.
- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.
- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).
- [x] Commit messages are formatted like
    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module
  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->

britter reviewed (commented):
This is my first contribution to the nix eco system. I'm still learning nix. Hope this makes sense!

liff commented:
Would it be possible to add another setting for customizing the dot directory? The environment variable `GRADLE_USER_HOME` can be used to change `~/.gradle` to something else. Something like `programs.gradle.home` perhaps, that sets `home.sessionVariables.GRADLE_USER_HOME` accordingly?

liff commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,106 @@
+      '';
+    };
+    gradleProperties = mkOption {
+      type = types.attrsOf types.str;
```
Would it be possible to use [RFC 0042-style](https://github.com/NixOS/rfcs/pull/42) settings here?

There is an example in NixOS’s [Apache Kafka module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/apache-kafka.nix#L37-L50).

britter replied:
Done!

britter commented:
@liff I'll have a look into both your suggestions. Thank you for the feedback!

liff commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,89 @@
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        The Gradle home directory.
```
Since this is managed by `home.file`, it’ll have to be a relative path:

```suggestion
        The Gradle home directory, relative to `home.homeDirectory`.
```

liff commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,87 @@
+    home.packages = [ cfg.package ];
+    home.file.".gradle/init.d" =
+      mkIf (cfg.initScripts != { }) { source = "${initScripts}/init.d"; };
+    home.sessionVariables =
```
You could maybe check the contents of `home-path/etc/profile.d/hm-session-vars.sh`? That’s what the [`sessionVariables` test does](https://github.com/nix-community/home-manager/blob/948703f3e71f1332a0cb535ebaf5cb14946e3724/tests/modules/home-environment/session-variables.nix).

britter commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,106 @@
+      '';
+    };
+    initScripts = mkOption {
+      type = types.attrsOf types.str;
```
Done

britter commented:
@liff this is now ready for the final review / merge. Thank you for your support. I learned a lot along the way. ❤️

liff commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+        [
+          {
+            name = "maven-local.gradle"
+            text = ${"''"}
```
`''` can be [quoted using a third `'`](https://nixos.org/manual/nix/stable/language/values.html?highlight=multi%20line#type-string).

```suggestion
            text = '''
```

britter commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+    gradleHome = "${config.home.homeDirectory}/${cfg.home}";
+  in mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+    home.file."${gradleHome}/init.d" =
```
You're right and this shows there's not test case covering this code path. I will add one.

britter replied:
Actually this should make the init script tests fail. 🤔

britter replied:
@liff I added a test for this explicitly, but it was green even with out this change 🤔 Maybe `home.file` detects this case and then removes `config.home.homeDirectory` from the path again? Anyways, I've applied your suggestion because I think it's easier to understand.

britter commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+                  mavenLocal()
+                }
+              }
+            ${"''"};
```
```suggestion
            ''';
```

britter commented:
@liff anything missing in order to get this merged?

liff commented:
Looks good to me, but I lack the power to merge 😔

britter commented:
@liff do you know by any chance who to ping to get this merged?

liff commented:
I guess it would be @rycee by default? Or you could ask on the [IRC channel](https://nix-community.github.io/home-manager/index.xhtml#preface).

It occurred to me that since this is a new module, perhaps it’s worthy of a [news entry](https://nix-community.github.io/home-manager/index.xhtml#sec-news). New modules usually seem to get one.

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+      '';
+    };
+    initScripts = mkOption {
+      type = types.either types.path (types.listOf initScript);
```
Perhaps `type = with types; attrsOf initScript`?

Would look something like

``` nix
initScripts = {
  "maven-local.gradle".text = "…";
  "another.init.gradle.kts".source = ./another.init.gradle.kts;
};
```

rycee replied:
To be clear, with the above I also suggest removing the possibility to specify just a path. Instead always require an attribute set. It will help future problems with merging. Specifically, it is unclear what

``` nix
mkMerge [
  { initScripts = ./path; }
  { initScripts."extra-init.gradle".source = ./path2; }
]
```

should do.

You could use `apply` to coerce `./path` to `{ source = ./path; }` but I generally think it is best to avoid `apply` to maintain predictability w.r.t merging.

britter replied:
Hey @rycee 👋🏻

My thinking was that if I had multiple init scripts, I'd like to reference a directory instead of listing them all. But I now understand that this doesn't work with merging configurations. I will change this.

britter replied:
Done

rycee commented on tests/modules/programs/gradle/alternate-home-settings.nix:
```diff
@@ -0,0 +1,25 @@
+    test.stubs.gradle = { };
+
+    nmt.script = ''
+      assertFileRegex home-path/etc/profile.d/hm-session-vars.sh 'export GRADLE_USER_HOME="/home/hm-user/.gbt"'
```
Perhaps

```suggestion
      assertFileContains home-path/etc/profile.d/hm-session-vars.sh \
        'export GRADLE_USER_HOME="/home/hm-user/.gbt"'
```

rycee commented on tests/modules/programs/gradle/example-settings.nix:
```diff
@@ -0,0 +1,33 @@
+      settings = {
+        "org.gradle.caching" = true;
+        "org.gradle.parallel" = true;
+        "org.gradle.java.home" = pkgs.jdk17;
```
Unfortunately you'll need to stub these jdk packages (or just use some other mocked packages), otherwise the test downloads will be enormous just to insert a short text into the output file.

rycee replied:
Had a look at this and unfortunately it is a bit tricky to accomplish this in this case. For now, I think the best possible solution is

``` patch
diff --git a/tests/modules/programs/gradle/example-settings.nix b/tests/modules/programs/gradle/example-settings.nix
index 36b9736f..19b8c070 100644
--- a/tests/modules/programs/gradle/example-settings.nix
+++ b/tests/modules/programs/gradle/example-settings.nix
@@ -14,17 +14,26 @@ with lib;
       };
     };
 
-    test.stubs.gradle = { };
+    programs.java.package =
+      pkgs.runCommandLocal "java" { home = ""; } "mkdir $out";
+
+    test.stubs = {
+      gradle = { };
+      jdk = { };
+      jdk8 = { };
+      jdk11 = { };
+      jdk17 = { };
+    };
 
     nmt.script = ''
       assertFileExists home-files/.gradle/gradle.properties
       assertFileContent home-files/.gradle/gradle.properties ${
-        pkgs.writeText "gradle.expected" ''
+        builtins.toFile "gradle.expected" ''
           # Generated with Nix
 
           org.gradle.caching = true
-          org.gradle.java.home = ${pkgs.jdk17}
-          org.gradle.java.installations.paths = ${pkgs.jdk8},${pkgs.jdk11}
+          org.gradle.java.home = @jdk17@
+          org.gradle.java.installations.paths = @jdk8@,@jdk11@
           org.gradle.parallel = true
         ''
       }
```

britter replied:
Done!

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+        '';
+      };
+      text = mkOption {
+        type = types.lines;
+        default =
+          if config.source != null then builtins.readFile config.source else "";
+        defaultText = literalExpression
+          "if source is defined, the content of source, otherwise empty";
+        description = ''
+          Text of the init script file.
+          If unset then the source option will be preferred.
+        '';
```
This is too complicated w.r.t merging and will cause confusion, and likely infinite recursion in some cases. I would suggest using a simpler setup similar to what is in `file-type.nix`:

```suggestion
        type = types.nullOr types.lines;
        default = null;
        description = ''
          Text of the init script file.
          If unset then the source option must be specified.
        '';
```

britter replied:
I adopted this from the [nushell module](https://github.com/nix-community/home-manager/blob/948703f3e71f1332a0cb535ebaf5cb14946e3724/modules/programs/nushell.nix#L64-L82) as suggested [here](https://github.com/nix-community/home-manager/pull/4667#discussion_r1413594398). Maybe that also needs to be fixed?

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+          If the text option is set, it will be preferred.
+        '';
+      };
+    };
```
Add `config` section where source is populated with the generated file, see `file-type.nix`.

rycee replied:
… when `mkIf (config.text != null)`, that is… 🙂

britter replied:
Wow, this was way more tricky than I anticipated. But it's done now.

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+        Key value pairs to write to gradle.properties in the Gradle home directory.
+      '';
+    };
+    initScripts = mkOption {
```
```suggestion

    initScripts = mkOption {
```

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+
+  options.programs.gradle = {
+    enable = mkEnableOption "Gradle Build Tool";
+    home = mkOption {
```
```suggestion

    home = mkOption {
```

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+      type = types.str;
+      default = defaultHomeDirectory;
+      description = ''
+        The Gradle home directory, relative to `home.homeDirectory`.
+
+        If set, the GRADLE_USER_HOME environment variable will be
+        set accordingly. Defaults to .gradle.
```
Can add some fancy markup.

```suggestion
        The Gradle home directory, relative to [](#opt-home.homeDirectory).

        If set, the {env}`GRADLE_USER_HOME` environment variable will be
        set accordingly. Defaults to {file}`.gradle`.
```

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+    settings = mkOption {
+      type = types.submodule { freeformType = settingsFormat.type; };
+      default = { };
+      example = {
```
I think this will look best as a literal expression:

```suggestion
      example = literalExpression ''
        {
```

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+    gradleHome = "${config.home.homeDirectory}/${cfg.home}";
+  in mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+    home.file."${cfg.home}/init.d" =
```
Can add an empty line between the bindings to make it less compact:
```suggestion

    home.file."${cfg.home}/init.d" =
```

rycee commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+      GRADLE_USER_HOME = gradleHome;
+    };
+    home.file."${cfg.home}/gradle.properties" = mkIf (cfg.settings != { }) {
+      source = (settingsFormat.generate "gradle.properties" cfg.settings);
```
```suggestion
      source = settingsFormat.generate "gradle.properties" cfg.settings;
```

rycee commented:
Thanks for the contribution! I've added a bunch of mostly stylistic comments. The most important is to reduce the test closure to avoid GBs of downloads whenever the tests run 🙂

britter commented on modules/programs/gradle.nix:
```diff
@@ -0,0 +1,134 @@
+      type = types.submodule { freeformType = settingsFormat.type; };
+      default = { };
+      example = {
+        "org.gradle.caching" = true;
+        "org.gradle.parallel" = true;
+        "org.gradle.jvmargs" = "-XX:MaxMetaspaceSize=384m";
+        "org.gradle.home" = pkgs.jdk17;
+      };
```
```suggestion
          "org.gradle.caching" = true;
          "org.gradle.parallel" = true;
          "org.gradle.jvmargs" = "-XX:MaxMetaspaceSize=384m";
          "org.gradle.home" = pkgs.jdk17;
        };
      ''
```

britter commented:
@rycee I think I addressed all your comments. Previously the code aggregated all init scripts in a store directory using `symlinkJoin`. I was unable to get that working with the new structure of the initScript submodule where the source is set from the text. So the current version links individual files from the store inside `.gradle/init.d`. Unfortunately I didn't find a more elegant solution than `home.file = mkMerge ...` which feels more complex than it should.

I also added a news entry.

rycee commented:
Thanks! Merged to master now 🙂
